src/economy/economy.cpp

```diff
@@ -3275,91 +5274,181 @@ void daily_update(sys::state& state, bool initiate_buildings) {
-					rd = std::max(rd - sp_pool, 0.0f);
-				}
-				state.world.commodity_set_global_market_pool(c, std::max(0.0f, wm_pool - rd));
+		if(state.world.commodity_get_money_rgo(cid)) {
+			return;
+		}
 
-				state.world.nation_set_imports(n, c, std::min(wm_pool, rd));
-			} else {
-				state.world.nation_set_imports(n, c, std::min(wm_pool, rd));
+		state.world.for_each_trade_route([&](auto trade_route) {
+			auto current_volume = state.world.trade_route_get_volume(trade_route, cid);
+			auto origin =
+				current_volume > 0.f
+				? state.world.trade_route_get_connected_markets(trade_route, 0)
+				: state.world.trade_route_get_connected_markets(trade_route, 1);
+			auto target =
+				current_volume <= 0.f
+				? state.world.trade_route_get_connected_markets(trade_route, 0)
+				: state.world.trade_route_get_connected_markets(trade_route, 1);
 
-				state.world.commodity_set_global_market_pool(c, std::max(0.0f, wm_pool - rd));
-				rd = std::max(rd - wm_pool, 0.0f);
+			auto trade_loss = 0.8f;
 
-				state.world.nation_set_domestic_market_pool(n, c, std::max(0.0f, dom_pool - rd));
-				rd = std::max(rd - dom_pool, 0.0f);
-				if(sl) {
-					state.world.nation_set_domestic_market_pool(sl, c, std::max(0.0f, sl_pool - rd));
-					rd = std::max(rd - sl_pool, 0.0f);
-				}
-				if(state.world.nation_get_drawing_on_stockpiles(n, c)) {
-					state.world.nation_set_stockpiles(n, c, std::max(0.0f, sp_pool - rd));
-				}
-			}
-		}
-	}
+			auto sat = state.world.market_get_direct_demand_satisfaction(origin, cid);
+
+			//reduce volume in case of low supply
+			current_volume = current_volume * std::max(0.99f, sat);
+			state.world.trade_route_set_volume(trade_route, cid, current_volume);
+
+			auto absolute_volume = std::abs(current_volume);
+
+			auto s_origin = state.world.market_get_zone_from_local_market(origin);
+			auto s_target = state.world.market_get_zone_from_local_market(target);
+
+			auto n_origin = state.world.state_instance_get_nation_from_state_ownership(s_origin);
+			auto n_target = state.world.state_instance_get_nation_from_state_ownership(s_target);
+
+			register_demand(state, origin, cid, absolute_volume, economy_reason::trade);
+		});
+	});
+
+	sanity_check(state);
 
 	/*
-	move remaining domestic supply to global pool, clear domestic market
+	perform actual consumption / purchasing subject to availability at markets:
 	*/
-	state.world.for_each_commodity([&](dcon::commodity_id c) {
-		// per good decay would be nice...
-		float decay = 0.5f;
-		float world_pool = state.world.commodity_get_global_market_pool(c) * decay;
-		ve::fp_vector sum;
-		state.world.execute_serial_over_nation([&](auto nids) {
-			sum = sum + state.world.nation_get_domestic_market_pool(nids, c);
-			state.world.nation_set_domestic_market_pool(nids, c, 0.0f);
-		});
-		state.world.commodity_set_global_market_pool(c, world_pool + sum.reduce());
+
+	state.world.execute_serial_over_market([&](auto ids) {
```

Comment:
I think that this could be done in parallel, or that it could be done in parallel with a few adjustments. If I am reading the code correctly, the only things that block parallelization are the effects on nation stockpiles/treasury that come from buying and selling from the stockpiles. And since that is only done in capital markets, and each nation has only one of those, it shouldn't be possible for this function to try to touch values for the same nation if it was run in parallel.

If my analysis is wrong here, it looks like it should still be relatively simple to parallelize. Simply create one or more buffers that are the size of the number of markets to record effects on the owning nation and instead of doing the addition/subtraction effect in function, write the change to the buffer. Then the changes can be applied in serial after the main work, done in parallel, has been finished.